File: src/main/java/net/explorviz/adapter/service/converter/SpanConverterImpl.java
package net.explorviz.adapter.service.converter;

import jakarta.enterprise.context.ApplicationScoped;
import net.explorviz.avro.Span;

/**
 * Converts a {@link io.opentelemetry.proto.trace.v1.Span} to a {@link Span}.
 */
@ApplicationScoped
public class SpanConverterImpl implements SpanConverter<Span> {

  private static final long TO_MILLISEC_DIVISOR = 1_000_000L;

  @Override
  public Span fromOpenTelemetrySpan(final io.opentelemetry.proto.trace.v1.Span ocSpan) {

    final AttributesReader attributesReader = new AttributesReader(ocSpan);

    String parentSpan = "";
    if (ocSpan.getParentSpanId().size() > 0) {
      parentSpan = IdHelper.converterSpanId(ocSpan.getParentSpanId().toByteArray());
    }

    final Span.Builder span = Span.newBuilder()
        .setLandscapeToken(attributesReader.getLandscapeToken())
        .setGitCommitChecksum(attributesReader.getGitCommitChecksum())
        .setParentSpanId(parentSpan)
        .setSpanId(IdHelper.converterSpanId(ocSpan.getSpanId().toByteArray()))
        .setTraceId(IdHelper.converterTraceId(ocSpan.getTraceId().toByteArray()))
        .setStartTimeEpochMilli(ocSpan.getStartTimeUnixNano() / TO_MILLISEC_DIVISOR)
        .setEndTimeEpochMilli(ocSpan.getEndTimeUnixNano() / TO_MILLISEC_DIVISOR);

    attributesReader.appendToSpan(span);

    return span.build();
  }
}

File: src/main/java/net/explorviz/adapter/service/converter/SpanConverterImpl.kt
package net.explorviz.adapter.service.converter

import jakarta.enterprise.context.ApplicationScoped
import net.explorviz.avro.Span

/** Converts a [io.opentelemetry.proto.trace.v1.Span] to a [Span]. */
@ApplicationScoped
class SpanConverterImpl : SpanConverter<Span> {
    override fun fromOpenTelemetrySpan(ocSpan: io.opentelemetry.proto.trace.v1.Span): Span {

        val attributesReader = AttributesReader(ocSpan)

        val parentSpan =
            if (ocSpan.parentSpanId.size() > 0) {
                IdHelper.convertSpanId(ocSpan.parentSpanId.toByteArray())
            } else {
                ""
            }

        val span =
            Span.newBuilder()
                .setLandscapeToken(attributesReader.landscapeToken)
                .setGitCommitChecksum(attributesReader.gitCommitChecksum)
                .setParentSpanId(parentSpan)
                .setSpanId(IdHelper.convertSpanId(ocSpan.spanId.toByteArray()))
                .setTraceId(IdHelper.convertTraceId(ocSpan.traceId.toByteArray()))
                .setStartTimeEpochMilli(ocSpan.startTimeUnixNano)
                .setEndTimeEpochMilli(ocSpan.endTimeUnixNano)

        attributesReader.appendToSpan(span)

        return span.build()
    }
}
